tools/integration_tests/concurrent_operations/concurrent_listing_test.go

```diff
@@ -33,20 +44,486 @@ func (s *concurrentListingTest) Setup(t *testing.T) {
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			f, err := os.Open(targetDir)
+			assert.Nil(t, err)
+
+			err = f.Close()
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Goroutine 1: Repeatedly calls Stat.
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			_, err := os.Stat(targetDir)
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Wait for goroutines or timeout.
+	done := make(chan bool, 1)
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+	select {
+	case <-done:
+		// Operation completed successfully before timeout.
+	case <-time.After(timeout):
+		assert.FailNow(t, "Possible deadlock")
+	}
+}
+
+// Test_Parallel_ReadDirAndLookUp tests for potential deadlocks or race conditions when
+// ReadDir() is called concurrently with LookUp of same dir.
+func (s *concurrentListingTest) Test_Parallel_ReadDirAndLookUp(t *testing.T) {
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_Parallel_ReadDirAndLookUp"
+	createDirectoryStructureForTestCase(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	wg.Add(2)
+	timeout := 200 * time.Second
+	iterationsPerGoroutine := 40
+
+	// Goroutine 1: Repeatedly calls Readdir
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			f, err := os.Open(targetDir)
+			assert.Nil(t, err)
+
+			_, err = f.Readdirnames(0)
+			assert.Nil(t, err)
+
+			err = f.Close()
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Goroutine 2: Repeatedly stats
+	go func() {
+		defer wg.Done()
+		for i := 0; i < iterationsPerGoroutine; i++ {
+			_, err := os.Stat(targetDir)
+			assert.Nil(t, err)
+		}
+	}()
+
+	// Wait for goroutines or timeout
+	done := make(chan bool, 1)
+	go func() {
+		wg.Wait()
+		done <- true
+	}()
+
+	select {
+	case <-done:
+		// Success: Both operations finished before timeout
+	case <-time.After(timeout):
+		assert.FailNow(t, "Possible deadlock or race condition detected during Readdir and directory operations")
+	}
+}
+
+// Test_MultipleConcurrentReadDir tests for potential deadlocks or race conditions
+// when multiple goroutines call Readdir() concurrently on the same directory.
+func (s *concurrentListingTest) Test_MultipleConcurrentReadDir(t *testing.T) {
+	t.Parallel() // Mark the test parallelizable.
+	testCaseDir := "Test_MultipleConcurrentReadDir"
+	createDirectoryStructureForTestCase(t, testCaseDir)
+	targetDir := path.Join(testDirPath, testCaseDir, "explicitDir")
+	var wg sync.WaitGroup
+	goroutineCount := 10          // Number of concurrent goroutines
+	iterationsPerGoroutine := 100 // Number of iterations per goroutine
+	wg.Add(goroutineCount)
+	timeout := 50 * time.Second
+
+	// Create multiple go routines to listing concurrently.
+	for i := 0; i < goroutineCount; i++ {
+		go func() {
+			defer wg.Done()
+
+			for j := 0; j < iterationsPerGoroutine; j++ {
+				f, err := os.Open(targetDir)
+				assert.Nil(t, err)
+
+				_, err = f.Readdirnames(-1) // Read all directory entries
```

Comment:
at some places we have used 0 is Readdirnames too. Let's use either -1 or 0 at all places